### ISSUE TITLE
Function returned from useMutation and useManualQuery is not memoized 

### ISSUE THREAD

**Issue by lynxtaa** [nearform/graphql-hooks#234]
### Package

`graphql-hooks`

### Environment

- `graphql-hooks` version: 3.5.0
- `react` version: 16.9.0-alpha.0
- Browser: Chrome

### Description

Function returned from useMutation and useManualQuery is not memoized and can't  be used in useEffect as a  dependency.

### How to reproduce

```jsx
const [shouldFetch, setShouldFetch] = useState(false)
const [fetchUser] = useManualQuery(GET_USER_QUERY)

useEffect(() => {
  if (shouldFetch) {
    fetchUser()
  }
}, [fetchUser, shouldFetch])

return <button onClick={() => setShouldFetch(true)} />
``` 

When button is clicked we expect only one request. But because fetchUser is not memoized internally, we get lots of requests and infinite render.

### Suggested solution (optional)

We can wrap fetchData in useCallback (see PR)

**Comment by stagedataloss** [comment]
test123test123

**Comment by lynxtaa** [comment]
As i found out you can use `client.request` from `ClientContext` (which doesn't change during rerenders) for manual queries and mutations in effects (Apollo has [similar approach](https://www.apollographql.com/docs/react/essentials/queries/#manually-firing-a-query)). Note, that caching won't work with `client.request`.

Example:
```jsx
import {ClientContext} from 'graphql-hooks'
...
const client = useContext(ClientContext)

useEffect() => {
  client.request({ query, variables })
}, [client])
```

**Comment by lynxtaa** [comment]
I think to achieve memoization without deep quallity checks of `options` object (which also may contain functions like `fetchOptionsOverrides` and `updateData`) we should omit the second argument in useMutation and useManualQuery. Then invalidating returned function based on query will be easy. Although it'll be breaking change. @bmullan91 what do you think? I can create PR based on this idea

**Comment by olistic** [comment]
@lynxtaa I find your example a bit contrived, to be honest. Is there a reason why you're using `useEffect` instead of calling `fetchUser` inside the `onClick` handler?

```js
const [fetchUser] = useManualQuery(GET_USER_QUERY)

return <button onClick={fetchUser} />
```

**Comment by lynxtaa** [comment]
@olistic sorry for the confusion, I oversimplified my example to highlight main issue. Problem is: we can't use useManualQuery and useMutation inside useEffect because it creates infinite render. Here are two more examples:

Using `useManualQuery`:
```jsx
function UserProfile({ id = null }) {
  const [getUser, { data, loading }] = useManualQuery(`
    query getUser($id: Int!) {
      user(id: $id) {
        login
      }
    }
  `)

  useEffect(() => {
    if (id) {
      getUser({ variables: { id } })
    }
  }, [id, getUser])

  return (
    <div>
      <h1>{id ? (loading ? 'Loading...' : data.login) : 'New User'}</h1>
    </div>
  )
}
```

Using `useMutation`
```jsx
function UserProfile() {
  const [data, setData] = useState('')
  const [savingStarted, setSavingStarted] = useState(false)
  const [saveData] = useMutation(`
    mutation saveData($data: String!) {
      saveData(data: $data) {
        id
      }
    }
  `)

  useEffect(() => {
    if (savingStarted) {
      let timer = setTimeout(
        () => saveData({ variables: { data } }).then(() => setSavingStarted(false)),
        5000,
      )
      return () => clearTimeout(timer)
    }
  }, [data, saveData, savingStarted])

  return (
    <div>
      <input
        value={data}
        onChange={event => setData(event.target.value)}
        readOnly={savingStarted}
      />
      <button type="button" onClick={() => setSavingStarted(!savingStarted)}>
        {savingStarted ? "I've changed my mind. Cancel" : 'Save'}
      </button>
    </div>
  )
}
```

Of course all of it can be solved using custom hook:
```jsx
function useUnchangingManualQuery(query) {
  const [fetch, result] = useManualQuery(query)
  const latestFetch = useRef(fetch)
  latestFetch.current = fetch
  return [latestFetch.current, result] 
}
```

But I think current behaviour is unintuitive and actually is a bug

**Comment by bmullan91** [comment]
Fixed in #298
